cluster-autoscaler/clusterstate/clusterstate.go

```diff
@@ -958,6 +984,44 @@ func getNotRegisteredNodes(allNodes []*apiv1.Node, cloudProviderNodeInstances ma
 	return notRegistered
 }
 
+// Calculates which of the registered nodes in Kubernetes that do not exist in cloud provider.
+func (csr *ClusterStateRegistry) getCloudProviderDeletedNodes(allNodes []*apiv1.Node, cloudProviderNodeInstances map[string][]cloudprovider.Instance) []*apiv1.Node {
+	nodesRemoved := make([]*apiv1.Node, 0)
+	currentCloudInstances := make(map[string]string, 0)
+	registeredNodes := make(map[string]*apiv1.Node, 0)
+	for nodeGroupName, instances := range cloudProviderNodeInstances {
+		for _, instance := range instances {
+			currentCloudInstances[instance.Id] = nodeGroupName
+		}
+	}
+	for _, node := range allNodes {
+		registeredNodes[node.Name] = node
+	}
+
+	// Fill previously deleted nodes, if they are still registered in Kubernetes
+	for nodeName, node := range csr.deletedNodes {
+		if _, found := registeredNodes[nodeName]; found {
+			nodesRemoved = append(nodesRemoved, node)
+		}
+	}
+
+	// Seek nodes that may have been deleted since last update
+	// cloudProviderNodeInstances are retrieved by nodeGroup,
+	// not autoscaled nodes will not be included
+	for _, instances := range csr.cloudProviderNodeInstances {
```

Comment:
Won't relying on previous state cause issues in case of CA restart? IIUC some nodes will be incorrectly considered as `Ready`/`Unready`/`NotStarted`, potentially leading to bad scaling decisions.

Reply:
That would definitely occur on a restart.  Would it make sense to create a new taint and add it onto the nodes?  There might be some additional overhead for the implementation, but should prevent that scenario.

I also thought about adding a check for each node in the `deletedNodes` (lines 1001 - 1006) to confirm it doesn't appear in the cloud provider nodes.  I don't have good insight to know if that is a case that could happen though.  I wouldn't want to keep flagging a node as deleted if the cloud provider node exists.

Reply:
That makes sense to me.  I agree that the API should eventually correct itself, and the scaling should only momentarily be impacted.  I've created a new commit that includes the backfill safety check mentioned and some other minor cosmetic code changes.